### PR TITLE
fix(risk-classifier): score test CI workflows below deployment (#1565)

### PR DESCRIPTION
**Pull request by lhpaul**
Closes #1565.

## What changed

`run-epic-risk-classifier.sh` treated every `.github/workflows/**` change as a sensitive category and escalated the PR to `high`. So a PR that wires a test suite into CI exceeded a `medium` ceiling and couldn't merge under delegated policy — **the risk model penalised closing a test-coverage gap**, which is exactly what #1537 asked for.

This wasn't hypothetical twice over: PR #1536, and then PR #1568 implementing #1537, which needed explicit owner authorization to merge for this reason alone.

Recognised test and lint workflows now score `medium` with their own reason string; deployment, release, and permission behavior still score `high`.

## Why filename, not content

The classifier is handed a list of changed **paths** and never file contents — `--input` carries `changed_files` and nothing else — so it cannot inspect what a workflow actually does.

That makes this deliberately an **allowlist that yields to a deny-list**:

- Allowed to `medium`: `test-*.yml`, `*-test(s).yml`, `*lint*.yml`, `shellcheck.yml`
- Forced to `high` regardless: any name containing `deploy`, `release`, `publish`, `tag`, `secret`, `credential`, `permission`, `policy`, `token`
- Anything unrecognised: stays `high`

Misjudging a deployment workflow as a test job is far worse than making someone name a test workflow conventionally. That trade is now documented for authors in `guardrails-enforcement.md`.

## Acceptance criteria

| AC | Result |
| -- | ------ |
| **AC-1** test/lint below deployment/release/permission | `medium` vs `high` |
| **AC-2** PR #1536's diff shape fits a `medium` ceiling | `medium`, `merge_permitted=true` |
| **AC-3** deployment change still `high` | `deploy.yml`, `auto-tag-release.yml` → `high` |
| **AC-4** self-describing-report interaction | Documented — see below |

Verified across the boundary cases:

| Change set | Risk |
| ---------- | ---- |
| `workflow-tests.yml` + `batch-merge.sh` + a test (the #1536/#1568 shape) | `medium` ✅ |
| `test-pr-review-loop.yml` alone | `medium` |
| `markdown-lint.yml`, `shellcheck.yml` | `medium` |
| `deploy.yml`, `auto-tag-release.yml`, `pr-policy.yml` | `high` |
| `e2e-regression.yml` (unrecognised) | `high` |
| `test-release.yml` (deny-list beats allowlist) | `high` |
| `workflow-tests.yml` + `auth/token.sh` | `high` |

## AC-4: documented, not guarded

A bug report must name its trigger to explain it, so **any keyword-driven gate in this framework mishandles its own bug reports** — which is precisely when correct handling matters most. You observed it twice on one run (the policy recommender on #1504, the overlap classifier on #1540).

I documented it as a known heuristic limit rather than building a guard, because a guard would need to know an item was filed *against* the heuristic it's about to run — that determination is the same keyword problem one level up. The guidance is to treat such a verdict as uninformative and read the item directly, and explicitly **not** to broaden the keyword list, which makes over-matching worse everywhere else.

If you'd rather have a mechanical guard (e.g. an opt-out label on the item), say so and I'll add it — I judged documentation the better trade, but it is a judgment call.

## Verification

- `test-run-epic-risk-classifier.sh`: **100 assertions passing**, 14 new
- ShellCheck clean; markdownlint, heuristic lint, CHANGELOG checks, snippet lint all clean
- **First real use of change-scoped CI** (merged in #1568): this change set selects **2 suites, not 66** — `test-run-epic-risk-classifier.sh` by naming convention and `test-may-merge-terminal-contract.sh` via its `# covers:` header on `guardrails-enforcement.md`. Both pass.

## Note on merging

Under its own new rule this PR is `medium` — it touches no workflow file at all, only the classifier and its tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved workflow risk classification: test and lint workflow changes now receive medium risk.
  * Deployment, release, publishing, security, permission, policy, and authentication-related workflow changes remain high risk.
  * Risk explanations now clearly reflect changed file paths and classification rules.

* **Documentation**
  * Documented filename-based classification limits and potential keyword-related false positives.

* **Tests**
  * Added coverage for workflow risk levels, merge thresholds, sensitive changes, and classification precedence.

### CODE REVIEW

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), coderabbit (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** coderabbit

**Advisory findings (non-blocking):**
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1569#issuecomment-5375651134))

*Posted automatically by `pr-review-loop.sh`.*

<details>
<summary>Reviewer-loop history (2 iterations)</summary>


```json
{
  "schema": "reviewer_loop_history.v1",
  "pr_number": 1569,
  "updated_at": "2026-08-21T21:47:46Z",
  "history_status": "available",
  "history_unavailable_reason": "",
  "entries": [
    {
      "iteration": 1,
      "recorded_at": "2026-08-21T21:38:36Z",
      "head_sha": "6efa9e68fbad9e7f68d90db6fc357d03e67be7ad",
      "run_id": "auto-1787348198-62893-20783",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 2,
      "recorded_at": "2026-08-21T21:47:46Z",
      "head_sha": "62fa6cd74badbb8f8bd74763416b5f927bb1b02e",
      "run_id": "auto-1787348749-3629-2871",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    }
  ]
}
```
</details>

File: CHANGELOG.md [hunk 1]
```diff
@@ -36,6 +36,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  ceiling and could not merge under delegated policy — the risk model penalised
+  closing a test-coverage gap, which is what #1537 asked for. Recognised test
+  and lint workflows (`test-*.yml`, `*-tests.yml`, `*lint*.yml`,
+  `shellcheck.yml`) now score `medium` with a distinct
+  `test or lint CI workflow change:` reason; deployment, release, and permission
+  behavior still score `high`. The classifier only ever receives changed paths,
+  never file contents, so this is judged by filename — deliberately an allowlist
+  that yields to a deny-list (`deploy`, `release`, `publish`, `tag`, `secret`,
+  `credential`, `permission`, `policy`, `token`), leaving an unrecognised
+  workflow at `high`.
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply CHANGELOG.md]: Applied in `62fa6cd74badbb8f8bd74763416b5f927bb1b02e`. You were right that the list was incomplete — the prose omitted `*-test.yml` and every `.yaml` variant, so an author using a valid name would have been surprised by a `high` classification. `guardrails-enforcement.md` now carries an authoritative table covering all four rules in match order (deny-list first), and the CHANGELOG keeps a summary and links to it instead of restating them.

File: scripts/development-workflow/run-epic-risk-classifier.sh [hunk 2]
```diff
@@ -316,6 +316,43 @@ base64_decode() {
...
+  case "$base" in
+    test-*.yml|test-*.yaml|\
+    *-test.yml|*-test.yaml|*-tests.yml|*-tests.yaml|\
+    *lint*.yml|*lint*.yaml|\
+    shellcheck.yml|shellcheck.yaml)
+      return 0
+      ;;
+  esac
+  return 1
+}
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/run-epic-risk-classifier.sh:356]: Applied in `62fa6cd74badbb8f8bd74763416b5f927bb1b02e`. `ci_workflow_is_test_or_lint` now requires exactly one non-empty path and calls `error_exit`; `ci_workflow_fixture`, `ci_workflow_risk`, and `ci_workflow_merge` validate their two arguments and exit 2 with a structured error. Verified the guard fires: calling the classifier helper with no argument prints `ERROR: ci_workflow_is_test_or_lint requires exactly one workflow path` and exits 2.

File: scripts/development-workflow/tests/test-run-epic-risk-classifier.sh [hunk 3]
```diff
@@ -356,6 +356,89 @@ medium_missing_output="$(classify_fixture "$medium_missing_fixture" medium)"
...
+
+# AC-3: deployment, release, and permission behavior stay high.
+run_test "ci_deploy_workflow_is_high" "high" \
+  "$(ci_workflow_risk wf-deploy '[".github/workflows/deploy.yml"]')"
+run_test "ci_deploy_workflow_blocked_at_medium" "false" \
+  "$(ci_workflow_merge wf-deploy-merge '[".github/workflows/deploy.yml"]')"
+run_test "ci_release_workflow_is_high" "high" \
+  "$(ci_workflow_risk wf-release '[".github/workflows/auto-tag-release.yml"]')"
+run_test "ci_policy_workflow_is_high" "high" \
+  "$(ci_workflow_risk wf-policy '[".github/workflows/pr-policy.yml"]')"
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/tests/test-run-epic-risk-classifier.sh:426]: Applied in `62fa6cd74badbb8f8bd74763416b5f927bb1b02e`, and extended a little further — the gap was wider than the one term. Added `permissions.yml`, `permission-sync.yml` (the singular), and `token-refresh.yml`, none of which the deny-list tests exercised. Also added `test-thing.yaml` and `smoke-test.yml` to cover the `.yaml` extension and the singular `*-test.yml` suffix, which had the same untested-pattern problem on the allowlist side. Suite is now 105 assertions.
